### PR TITLE
Avoid race condition in `GC_Example` test

### LINKED ISSUE #10643: Jobs appear to be stuck on Base_Tests
https://github.com/enso-org/enso/actions/runs/10055728974/job/27793016991

Test temporarily disabled in https://github.com/enso-org/enso/pull/10642.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Re-enables https://github.com/enso-org/enso/pull/10602/files#r1690919173 - uses `IO.println` to allow us to see what the CI actually does.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Unit tests are being analyzed

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
First run:
- [linux is fine](https://github.com/enso-org/enso/actions/runs/10089668131/job/27897556726?pr=10665#step:7:1145)
- [mac is fine](https://github.com/enso-org/enso/actions/runs/10089668131/job/27897557101?pr=10665#step:7:2049)
- [windows is fine](https://github.com/enso-org/enso/actions/runs/10089668131/job/27897557511?pr=10665#step:7:1108)

OK, let's try to schedule another run.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java:186]: To continue the analysis at [previous PR](https://github.com/enso-org/enso/pull/10602/files#r1690926438):

> For recurring actions ... [Future.isDone()](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/Future.html#isDone()) will return true only if the action was canceled.

Calling `request.cancel(false)` in the `synchronized (pendingItems)` section and then checking if `isCancelled()` when other threads arrive.

**Comment by JaroslavTulach** [comment]
Another successful run:
- [correct on Linux](https://github.com/enso-org/enso/actions/runs/10091359395/job/27902769321?pr=10665#step:7:1544)
- [correct on Mac](https://github.com/enso-org/enso/actions/runs/10091359395/job/27902769711?pr=10665#step:7:2048)
- [correct on Windows](https://github.com/enso-org/enso/actions/runs/10091359395/job/27902770067?pr=10665#step:7:1118)

What's next?

**Comment by JaroslavTulach** [comment]
We might [hit the GC problem](https://github.com/enso-org/enso/actions/runs/10091359395/job/27902766177?pr=10665) with Native Image tests:
```
DEBUG main_internal:Running process for the output.{command=Command:
	ENSO_DATA_DIRECTORY="/Users/runner/work/enso/enso/built-distribution/enso-engine-2024.3.1-dev-macos-amd64/enso-2024.3.1-dev" "/Users/runner/work/enso/enso/built-distribution/enso-engine-2024.3.1-dev-macos-amd64/enso-2024.3.1-dev/bin/enso" "--run" "/Users/runner/work/enso/enso/test/Base_Tests"
	with environment overrides:
		ENSO_DATA_DIRECTORY=/Users/runner/work/enso/enso/built-distribution/enso-engine-2024.3.1-dev-macos-amd64/enso-2024.3.1-dev}: ide_ci::program::command: Spawning.
```
I guess I should enable log output, at least.

**Review by GregoryTravis — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
All checks have passed including the [native image run on Linux with output](https://github.com/enso-org/enso/actions/runs/10096990465/job/27920785535?pr=10665#step:7:3770).

**Comment by JaroslavTulach** [comment]
And we have a [test failure on Linux](https://github.com/enso-org/enso/actions/runs/10096990465/job/27948766195?pr=10665#step:7:3585):
```
 INFO ide_ci::program::command: enso ℹ️ Still 10418 resources to clean up...
 INFO ide_ci::program::command: enso ℹ️ Still 10418 resources to clean up...
 INFO ide_ci::program::command: enso ℹ️ Still 10418 resources to clean up...
 INFO ide_ci::program::command: enso ℹ️ Still 10418 resources to clean up...
```
I am inclined to believe the test is wrong...

File: test/Base_Tests/src/Runtime/GC_Example.enso
Comment (JaroslavTulach) [inline test/Base_Tests/src/Runtime/GC_Example.enso:16]: @radeusgd, I am inclined to believe the [test is wrong](https://github.com/enso-org/enso/pull/10665#issuecomment-2252134656). The test calls:
```ruby
counter.modify (+1)
```
or `x-> x-1`. However those are not **atomic operations** any may get interrupted by the finalizer after the `x` value is read, but before its updated value is written. If another few finalizers are executed at that moment, then the counter will be off.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Hmmm that may be true

Comment (JaroslavTulach) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: We can trivially fix the test, if we allocate the resources first and only later clear them. Or if each resource gets its own `Ref` to decrement.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: How will that help? We will not get clashes between `+1` and `-1` but we can still get problems due to non-atomic `+1` (and `-1` respectively) in each batch.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: I mean this problem among even uniform increments: https://github.com/openjdk/jcstress/blob/master/jcstress-samples/src/main/java/org/openjdk/jcstress/samples/api/API_01_Simple.java

i.e. if there's `N` actors concurrently incrementing non-atomically, we don't have the guarantee that all `N` increments will be registered - so we will still get the test stuck.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: What if we do the following:
1. Introduce a timeout - if the cleanup did not succeed after ~5 minutes - let's make that a test failure - this will avoid any potential problem from blocking the runner for several hours, and it will actually make the tests report come up and be more helpful.
2. Maybe add some margins for the test to finish? I.e. do it in phases like you suggest, check that e.g. 100 000 resources caused the counter to increment to at least 90 000, and then check that after cleanup it went below 10 000 (and be OK with it getting below 0).

(1) will ensure we are not stuck and (2) should make it far less likely to fail the test due to concurrent behaviour, while still testing the core of this issue - multiple finalizers are scheduled at once.

Comment (JaroslavTulach) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Solved in 3c93430 by holding all the resources in a `Vector` and releasing them only after 3rd round of `Runtime.gc`.

Comment (JaroslavTulach) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: > I mean this problem among even uniform increments: https://github.com/openjdk/jcstress/blob/master/jcstress-samples/src/main/java/org/openjdk/jcstress/samples/api/API_01_Simple.java
> 
> i.e. if there's `N` actors concurrently incrementing non-atomically, we don't have the guarantee that all `N` increments will be registered - so we will still get the test stuck.

Please note that the test is running completely _single threaded_. It just schedules _interrupt_ (e.g. _safepoint action_ with a finalizer) at a random moment. 

If the _finalizer_ runs while we are still allocating resources, then we may have two methods on stack that manipulate the `Ref` `counter` and their code is certainly not ready for re-entrance. However if we first allocate everything and only then start running finalizers, then there always be only a single method on stack manipulating the counter - as we disallow _nesting of finalizer interrupts_  since #10602 and there is no longer any other method manipulating the `counter` on stack.

Unless I am mistaken the failure has always happened in _native image_ compiled `enso` launcher. Probably because the Truffle compilation there happens more quickly than in JVM and the code was still allocating... while in the JVM the compilation happens later when all the resources are already allocated. Since 3c93430 this difference shall be suppressed.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Ah, right! Thanks for the explanation, I forgot that we now disallow finalizers from interrupting each other, so indeed there is no place for unexpected races like that. I was wrongly viewing it as multithreaded

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Why do we need the 3 iterations of `Runtime.gc` before we release the resources?

Intuitively it feels like it should be enough to release the resources right away after they all were created. I don't understand what point these 3 iterations of GC server.

File: test/Base_Tests/src/Runtime/GC_Example.enso
Comment (radeusgd) [inline test/Base_Tests/src/Runtime/GC_Example.enso]: This will mean that this message is printed every 100 iterations. But it suffices to do this only once. Can we remove the `% 100`?

```suggestion
            if i == 3 then
                println "Releasing holder with resources: "+holder.get.length.to_text
                holder.put []
```

File: test/Base_Tests/src/Runtime/GC_Example.enso
Comment (JaroslavTulach) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Do you want to rewrite the `GC_Example` code? It is your code from #10211 anyway. I am just trying to make it reliable.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: It feels like there's not that much to change, but sure - I can try.

But I need to understand - was there any reason why we'd call the 3 GC rounds before starting the cleanup? From my perspective there is no reason so I'd change to start GCing only afterwards, but I'm afraid I may be missing something and in such case my rewrite could break stuff again.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Also if I understand your explanation correctly - my suggestion with 'margins' is not really needed because as it is running on a single thread, we should always get back to `0` after cleanup is done, right?

So the only meaningful change would be to add the timeout?

Comment (JaroslavTulach) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: >  suggestion with 'margins' is not really needed

The whole point of the test was to verify we always get the _exact same number_ of allocations as well as finalizer calls. "margins" just defeat the purpose of the test, I'd say.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Yes, I think you are right.

Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: I've modified the test to fail if a timeout of 5 minutes is exceeded:
```
Allocating 100000 resources...ocate lots of resources at once
Cleaning up 100000 resources
Still 368 resources to clean up...
Still 368 resources to clean up...
Still 368 resources to clean up...
Still 368 resources to clean up...
Still 368 resources to clean up...
Timeout has been reached, but there are 368 resources remaining that were not cleaned up.
[FAILED] Managed_Resource: [3/4, 60348ms]

    - [FAILED] allocate lots of resources at once [60222ms]
        Reason: An unexpected panic was thrown: (Illegal_State.Error 'Timeout has been reached, but there are 368 resources remaining that were not cleaned up.' Nothing)
        at <enso> Panic.throw(Internal)
```

Comment (JaroslavTulach) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Five minutes timeout is a bit too excessive for something that runs in 5s, but we don't expect the test to fail anymore.

**Comment by JaroslavTulach** [comment]
Previous runs blocked by
- #10695

let's hope we merge by end of this week

File: test/Base_Tests/src/Runtime/GC_Example.enso
Comment (radeusgd) [reply test/Base_Tests/src/Runtime/GC_Example.enso:16]: Fair, I wanted to have a larger timeout to avoid spurious failures under load, 5 mins is still better than 6 hours. But if you think we should change I'm happy to change it e.g. to 1 min or even 30s.

File: build/build/src/engine/context.rs
Comment (Frizi) [inline build/build/src/engine/context.rs:1]: Are there any test runs that failed "successfuly", to demonstrate that this code will properly flag failures? 

Normally my assumption would be that the stdout checks are here due to the test runner not returning appropriate exit codes on failures. But this code always looked like this, so maybe the check was superfluous. Either way, this need to be verified to not report false positives on CI before it is merged.

**Review by Frizi — APPROVED** [review]

File: build/build/src/engine/context.rs
Comment (JaroslavTulach) [reply build/build/src/engine/context.rs:1]: Thank you Pawel for your review. There [was a failure](https://github.com/enso-org/enso/pull/10665#issuecomment-2252134656) so we know the output is printed, but there was no failure - the execution timed out only. Let's check the [behavior with an artificial failure](https://github.com/enso-org/enso/pull/10665/commits/8875dc1de54bd844f85de529001b3b91e714c5ff):
- [failure on Linux](https://github.com/enso-org/enso/actions/runs/10139825671/job/28033833821?pr=10665#step:7:3764)
- [failure on Windows](https://github.com/enso-org/enso/actions/runs/10139825671/job/28033837469?pr=10665#step:7:1763)
- [failure on Mac](https://github.com/enso-org/enso/actions/runs/10139825671/job/28033834669?pr=10665#step:7:2688)

Let's back the commit out.

**Review by hubertp — APPROVED** [review]

**Pull request merged**
